failure to recover from backup v2.1.1

I've been doing a bit of fault injection for etcd-mesos, and a feature I'm testing is for cluster re-seeding when a majority of nodes are killed, and a backup-restore needs to occur.  I'm running `etcdctl backup` on the server while simultaneously killing the running server (the admin guide only says that windows users need to stop fully first), and restarting the server with `--force-new-cluster` afterward using the result of etcdctl backup.  It leads to this crash:
```
2015/08/27 15:34:10 etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 4
2015/08/27 15:34:10 etcdmain: the server is already initialized as member before, starting as etcd member...
2015/08/27 15:34:10 etcdmain: listening for peers on http://localhost.localdomain:30000
2015/08/27 15:34:10 etcdmain: listening for client requests on http://localhost.localdomain:30001
2015/08/27 15:34:10 etcdserver: recovered store from snapshot at index 1410334
2015/08/27 15:34:10 etcdserver: name = etcd-1440711567
2015/08/27 15:34:10 etcdserver: force new cluster
2015/08/27 15:34:10 etcdserver: data dir = etcd_data
2015/08/27 15:34:10 etcdserver: member dir = etcd_data/member
2015/08/27 15:34:10 etcdserver: heartbeat = 100ms
2015/08/27 15:34:10 etcdserver: election = 1000ms
2015/08/27 15:34:10 etcdserver: snapshot count = 10000
2015/08/27 15:34:10 etcdserver: advertise client URLs = http://localhost.localdomain:30001
2015/08/27 15:34:10 etcdserver: loaded cluster information from store: <nil>
2015/08/27 15:34:10 etcdserver: ConfChange Type should be either ConfChangeAddNode or ConfChangeRemoveNode!
panic: ConfChange Type should be either ConfChangeAddNode or ConfChangeRemoveNode!

goroutine 1 [running]:
github.com/coreos/etcd/Godeps/_workspace/src/github.com/coreos/pkg/capnslog.(*PackageLogger).Panicf(0xc2080a91a0, 0xa5d5f0, 0x4b, 0x0, 0x0, 0x0)
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/coreos/pkg/capnslog/pkg_logger.go:73 +0x185
github.com/coreos/etcd/etcdserver.getIDs(0xc20800a240, 0xc2095dc000, 0x19db, 0x1eaa, 0x0, 0x0, 0x0)
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:341 +0x53f
github.com/coreos/etcd/etcdserver.restartAsStandaloneNode(0xc2080ac000, 0xc20800a240, 0x29, 0xc208db1210, 0x0, 0x0, 0x0, 0x0)
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:283 +0x30d
github.com/coreos/etcd/etcdserver.NewServer(0xc2080ac000, 0xc20800a480, 0x0, 0x0)
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:280 +0x2a49
github.com/coreos/etcd/etcdmain.startEtcd(0xc208068d80, 0x0, 0x0, 0x0)
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:203 +0x1446
github.com/coreos/etcd/etcdmain.Main()
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:100 +0x8b2
main.main()
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/main.go:29 +0x1b

goroutine 13 [syscall]:
os/signal.loop()
	/usr/lib/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.initÂ·1
	/usr/lib/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 19 [chan receive]:
github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog.(*loggingT).flushDaemon(0xc48ee0)
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x78
created by github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog.initÂ·1
	/home/tyler/src/etcd-mesos/_vendor/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2a7
```

A copy of the data directory and a small invocation script to trigger the bug is located here: https://github.com/spacejam/etcd-bug-data

--- a/etcdserver/raft.go
+++ b/etcdserver/raft.go
@@ -382,6 +382,8 @@ func getIDs(snap *raftpb.Snapshot, ents []raftpb.Entry) []uint64 {
 			ids[cc.NodeID] = true
 		case raftpb.ConfChangeRemoveNode:
 			delete(ids, cc.NodeID)
+		case raftpb.ConfChangeUpdateNode:
+			// do nothing
 		default:
 			plog.Panicf("ConfChange Type should be either ConfChangeAddNode or ConfChangeRemoveNode!")
 		}

--- a/etcdserver/raft_test.go
+++ b/etcdserver/raft_test.go
@@ -32,6 +32,8 @@ func TestGetIDs(t *testing.T) {
 	removecc := &raftpb.ConfChange{Type: raftpb.ConfChangeRemoveNode, NodeID: 2}
 	removeEntry := raftpb.Entry{Type: raftpb.EntryConfChange, Data: pbutil.MustMarshal(removecc)}
 	normalEntry := raftpb.Entry{Type: raftpb.EntryNormal}
+	updatecc := &raftpb.ConfChange{Type: raftpb.ConfChangeUpdateNode, NodeID: 2}
+	updateEntry := raftpb.Entry{Type: raftpb.EntryConfChange, Data: pbutil.MustMarshal(updatecc)}
 
 	tests := []struct {
 		confState *raftpb.ConfState
@@ -48,6 +50,8 @@ func TestGetIDs(t *testing.T) {
 			[]raftpb.Entry{addEntry, removeEntry}, []uint64{1}},
 		{&raftpb.ConfState{Nodes: []uint64{1}},
 			[]raftpb.Entry{addEntry, normalEntry}, []uint64{1, 2}},
+		{&raftpb.ConfState{Nodes: []uint64{1}},
+			[]raftpb.Entry{addEntry, normalEntry, updateEntry}, []uint64{1, 2}},
 		{&raftpb.ConfState{Nodes: []uint64{1}},
 			[]raftpb.Entry{addEntry, removeEntry, normalEntry}, []uint64{1}},
 	}

